src/modules/ekf2/ekf2_main.cpp

```diff
@@ -1144,39 +1147,52 @@ void Ekf2::run()
 
 		// get external vision data
 		// if error estimates are unavailable, use parameter defined defaults
-		bool vision_position_updated = false;
-		bool vision_attitude_updated = false;
-		orb_check(_ev_pos_sub, &vision_position_updated);
-		orb_check(_ev_att_sub, &vision_attitude_updated);
-
-		if (vision_position_updated || vision_attitude_updated) {
-			// copy both attitude & position if either updated, we need both to fill a single ext_vision_message
-			vehicle_attitude_s ev_att = {};
-			orb_copy(ORB_ID(vehicle_vision_attitude), _ev_att_sub, &ev_att);
+		bool visual_odometry_updated = false;
+		orb_check(_ev_odom_sub, &visual_odometry_updated);
 
-			vehicle_local_position_s ev_pos = {};
-			orb_copy(ORB_ID(vehicle_vision_position), _ev_pos_sub, &ev_pos);
+		if (visual_odometry_updated) {
+			// copy both attitude & position, we need both to fill a single ext_vision_message
+			vehicle_odometry_s ev_odom;
+			orb_copy(ORB_ID(vehicle_visual_odometry), _ev_odom_sub, &ev_odom);
 
 			ext_vision_message ev_data;
-			ev_data.posNED(0) = ev_pos.x;
-			ev_data.posNED(1) = ev_pos.y;
-			ev_data.posNED(2) = ev_pos.z;
-			matrix::Quatf q(ev_att.q);
-			ev_data.quat = q;
-
-			// position measurement error from parameters. TODO : use covariances from topic
-			ev_data.posErr = fmaxf(_ev_pos_noise.get(), fmaxf(ev_pos.eph, ev_pos.epv));
-			ev_data.angErr = _ev_ang_noise.get();
-
-			// only set data if all positions and velocities are valid
-			if (ev_pos.xy_valid && ev_pos.z_valid && ev_pos.v_xy_valid && ev_pos.v_z_valid) {
+
+			// check for valid position data
+			if (PX4_ISFINITE(ev_odom.x) && PX4_ISFINITE(ev_odom.y) && PX4_ISFINITE(ev_odom.z)) {
+				ev_data.posNED(0) = ev_odom.x;
+				ev_data.posNED(1) = ev_odom.y;
+				ev_data.posNED(2) = ev_odom.z;
+
+				// position measurement error from parameters
+				if (PX4_ISFINITE(ev_odom.pose_covariance[0])) {
+					ev_data.posErr = fmaxf(_ev_pos_noise.get(), sqrtf(fmaxf(ev_odom.pose_covariance[0], ev_odom.pose_covariance[6])));
+
+				} else {
+					ev_data.posErr = _ev_pos_noise.get();
+				}
+			}
+
+			// check for valid orientation data
+			if (PX4_ISFINITE(ev_odom.q[0])) {
+				ev_data.quat = matrix::Quatf(ev_odom.q);
+
+				// orientation measurement error from parameters
+				if (PX4_ISFINITE(ev_odom.pose_covariance[15])) {
+					ev_data.angErr = fmaxf(_ev_ang_noise.get(), sqrtf(fmaxf(ev_odom.pose_covariance[15], fmaxf(ev_odom.pose_covariance[18],
+							       ev_odom.pose_covariance[20]))));
+
+				} else {
+					ev_data.angErr = _ev_ang_noise.get();
+				}
+			}
+
+			// only set data if all positions and orientation are valid
+			if (ev_data.posErr < ep_max_std_dev && ev_data.angErr < eo_max_std_dev) {
 				// use timestamp from external computer, clocks are synchronized when using MAVROS
-				_ekf.setExtVisionData(vision_position_updated ? ev_pos.timestamp : ev_att.timestamp, &ev_data);
+				_ekf.setExtVisionData(ev_odom.timestamp, &ev_data);
```

Comment:
Looks like position and orientation are copied into `ev_data` and sent to `_ekf` but linear and angular velocities are ignored?

Reply:
EKF2 doesn't currently handle those.